docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+#### Pipeline Step Composition
+
+Pipeline steps can be composed, and return a reference to the composed pipeline step. For example, if you wanted to create a pipeline that chunked a document and then embedded the chunks, you could do something like this:
+
+```python
+to_add = Embed(Chunk(data))
+```
+
+Or perhaps you want to chunk, embed and then quantize the embeddings:
+
+```python
+to_add = Quantize(Embed(Chunk(data)))
+```
+
+NOTE: This syntax could be confusing as the order of operations is reversed. We could consider alternative designs.
+
+#### Default Pipeline and Pipeline Persistence
+
+It is important to preserve the existing behavior of our API where users can simply pass text to the add step and have it be embedded. To do this, we will create a default pipeline that is used when no pipeline is provided. This default pipeline will simply embed the text. This default pipeline will be persisted with the collection, and will be used when no pipeline is provided. This means that users can create a collection with a default pipeline, and then query the collection without having to provide the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
+
+Pipelines can be persisted by serializing the pipeline steps and storing them in the metadata of the collection. This means that when a collection is loaded, the pipeline steps can be deserialized and used to recreate the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
+
+For example, the default pipeline would simply Embed the text, and would be persisted as follows:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+
+# ... some time later
+
+collection = client.get_collection("my_collection") # The default pipeline is automatically loaded
+```
+
+#### Query Pipelines
+
+In order to support embedding the data automatically for the user on query, we propose the introduction of the query_pipeline, which is the default pipeline run on the query data. This pipeline is persisted with the collection, and is used when no pipeline is provided. This means that users can create a collection with a default query pipeline, and then query the collection without having to provide the query pipeline.
+
+For example, the default query pipeline would simply Embed the text, and would be defined as follows:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(), query_pipeline=DefaultEmbeddingFunctionPipelineStep()) # The defaults would not actually have to be specified, as they would be the default values. Here we are specifying them for clarity.
```

Comment:
Haha, I like the "inversion of control". With this, LC becomes an upstream dep of Chroma